compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -972,7 +984,21 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
           paramProxy.get(tree.tpe) match {
             case Some(t) if tree.isTerm && t.isSingleton =>
               val inlinedSingleton = singleton(t).withSpan(argSpan)
-              inlinedFromOutside(inlinedSingleton)(tree.span)
+              val tree1 = inlinedFromOutside(inlinedSingleton)(tree.span)
+              if defn.isTupleNType(tree.tpe.widenTermRefExpr) && !defn.isTupleNType(t.widenTermRefExpr) then
+                // See tests/pos/i14182b.scala
+                // FIXME: We have a `tree1` that has a term ref `tup.type` with underlying `1 *: 2 *: EmptyTuple` type but the original reference is to a `Tuple2[Int, Int]`
+                // If we take the `tree1` then it is impossible to select the `_1` field. If we take the parameter type we can lose singletons references that we should keep.
+                // Ideally we should ascribe it to a `tup.type & Tuple2[Int, Int]` but that is not working because we have the some special sub-typing rule that makes
+                // `1 *: 2 *: EmptyTuple <:< Tuple2[Int, Int]`. Therefore the glb of the two types ends up being `tup.type` which drops the extra information we need to add.
+                // None of the following attempts work.
```

Comment:
@odersky do you have any idea how we could make this work?

Reply:
An alternative would be to teach member resolution that if we have a generic tuple with a known size that erases to `TupleN` it has the members of that `TupleN` class. For example a `Int *: EmptyTuple` would have the member `_1` because it has the members of `TupleN`.

Reply:
This member selection issue is not contained in inlining. It was fairly straightforward to reproduce using type parameters in https://github.com/lampepfl/dotty/issues/14215.

Reply:
> @odersky said to look at
> Line 3848 Typer:       pt match
>       case pt: SelectionProto =>
> insert type ascription and readapt